### PR TITLE
Add range iterator and contains_any/all to all container types

### LINKED ISSUE #23: Missing contains_keys_in_range and get_in_range
Hello,

I've just discovered your crate and it suits all my needs but one. I need to know whether the maps (in my case) contains any key in a provided range.

For example:

```rust
let mut map = RangeMap::new();

// Fill map.
map.insert(10..20, "a");
map.insert(50..70, "b");

// Checks.
map.contains_keys_in_range(5..9); // false
map.contains_keys_in_range(5..35); // true
map.contains_kesy_in_range(20..45); // false
```

Also, it would be very useful to retrieve all elements that collide with the range:

```rust
let mut map = RangeMap::new();

// Fill map.
map.insert(10..20, "a");
map.insert(50..70, "b");

// Get.
map.get_in_range(5..9); // []
map.get_in_range(5..35); // ["a"]
map.get_in_range(20..45); // []
```

> Note: feel free to change the method names.

### PR DESCRIPTION
**Pull request by Rua**
Fixes #23. The implementation mostly reuses existing patterns and code in the crate already.

### CODE REVIEW

**Comment by jeffparsons** [comment]
Thanks for this @Rua, and sorry about the delay in getting to it.

I'm concerned this might miss stored ranges that straddle the start of the provided range. I've been meaning to make the tests in this crate a bit friendlier to work with, so I think I'll do that before merging your PR — then it'll be easier to add tests for this behavior!

**Comment by jeffparsons** [comment]
Thanks for this @Rua, and sorry about the delay in getting to it.

I'm concerned this might miss stored ranges that straddle the start of the provided range. I've been meaning to make the tests in this crate a bit friendlier to work with, so I think I'll do that before merging your PR — then it'll be easier to add tests for this behavior!

File: src/inclusive_map.rs [hunk 1]
```diff
@@ -427,6 +439,28 @@ where
...
+        use core::ops::Bound;
+
+        let start = self
+            .get_key_value(range.start())
+            .map_or(range.start(), |(k, _v)| k.start());
+        let end = range.end();
+
+        RangeIter {
+            inner: self.btm.range((
+                Bound::Included(RangeInclusiveStartWrapper::new(
```
Comment (jeffparsons) [inline src/inclusive_map.rs:456]: This is the bit that looks suspicious to me: what happens if a stored range starts before the start of the given `range`? (I'll add a test to cover this after sprucing up the existing test patterns a bit.)

Comment (jeffparsons) [inline src/inclusive_map.rs:456]: This is the bit that looks suspicious to me: what happens if a stored range starts before the start of the given `range`? (I'll add a test to cover this after sprucing up the existing test patterns a bit.)

**Comment by rumpuslabs** [comment]
This is my main use case for this crate, what will it take to get this merged? Can I help?

**Comment by jeffparsons** [comment]
> This is my main use case for this crate, what will it take to get this merged? Can I help?

@rumpuslabs Yes, but probably not promptly due to a general lack of free time.

Can you tell me a bit more about your use case? I've been thinking a bit about what I'd like these APIs to look like, so hearing more about how people hope to use them might help.

**Comment by rumpuslabs** [comment]
I have a set of objects that take up ranges of an address space, and I want to know when adding a new object to the set if it overlaps with any of the existing ones (and if so, which ones so that I can generate a useful error message). So just being able to iterate over all the entries in the RangeMap that overlap with a given range would be sufficient for me, since if the iterator is empty I would generate no errors. It would look something like:

```rust
let mut errors = Vec::new();
for (_, conflicting_device) in rangemap.overlaps(new_device.range) {
    errors.push(Error::OverlappingAddresses(new_device, conflicting_device))
}
```

This PR called it just `range()` but I've used `overlaps()` since it seems to make a nice counterpart to `gaps()`.

**Comment by jeffparsons** [comment]
Do you imagine the iterator yielding whatever ranges are stored, or ranges truncated as necessary to fit the query range? E.g. if you have a map/set containing `[5..=10, 20..=25]`, and you query for ranges overlapping `15..=22`, would you expect to get `[20..=25]` or `[20..=22]`?

The former ("all stored ranges overlapping this given range") is easier to implement and implies less copying/cloning that the user might not have intended (because you can just return borrowed ranges, like in `iter()`), but I imagine some users might expect the latter. The latter could easily be implemented in terms of the former... but I'm not sure whether I'd bother implementing it at all or just leave that as an exercise to the reader if it's pretty rarely desired. :man_shrugging:

**Comment by rumpuslabs** [comment]
As you say, the truncated version can be implemented in terms of the easier to implement version, by doing a range union operation on the input and output ranges, but not the other way around since you're throwing away information by doing the truncation.

**Comment by jeffparsons** [comment]
Okay, thanks for that feedback.

There's a chance this is really easy (if `Borrow` behaves the way I think/hope), in which case I may be able to do it this week. Otherwise it's much fiddlier. I'll take a swing and let you know how it goes.

**Comment by jeffparsons** [comment]
Closing this now that https://github.com/jeffparsons/rangemap/pull/59 and https://github.com/jeffparsons/rangemap/pull/61 have been merged.

**Pull request closed without merging**